plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,457 @@
+	page, err := hypervisors.List(o.compute).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list hypervisors: %v", err)
+	}
+	hypervisors, err := hypervisors.ExtractHypervisors(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract hypervisors: %v", err)
+	}
+	for _, hypervisor := range hypervisors {
+		o.hypervisors[hypervisor.ID] = hypervisor
+	}
+	return nil
+}
+
+// gatherFlavors collects flavors from the OpenStack API.
+func (o *OpenStack) gatherFlavors() error {
+	page, err := flavors.ListDetail(o.compute, &flavors.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list flavors: %v", err)
+	}
+	flavors, err := flavors.ExtractFlavors(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract flavors: %v", err)
+	}
+	for _, flavor := range flavors {
+		o.flavors[flavor.ID] = flavor
+	}
+	return nil
+}
+
+// gatherServers collects servers from the OpenStack API.
+func (o *OpenStack) gatherServers() error {
+	page, err := servers.List(o.compute, &servers.ListOpts{AllTenants: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list servers: %v", err)
+	}
+	servers, err := servers.ExtractServers(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract servers: %v", err)
+	}
+	for _, server := range servers {
+		o.servers[server.ID] = server
+	}
+	return nil
+}
+
+// gatherVolumes collects volumes from the OpenStack API.
+func (o *OpenStack) gatherVolumes() error {
+	if !o.services.ContainsService(volumeV2Service) {
+		return nil
+	}
+	page, err := volumes.List(o.volume, &volumes.ListOpts{AllTenants: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list volumes: %v", err)
+	}
+	v := []volume{}
+	volumes.ExtractVolumesInto(page, &v)
+	for _, volume := range v {
+		o.volumes[volume.ID] = volume
+	}
+	return nil
+}
+
+// gatherStoragePools collects storage pools from the OpenStack API.
+func (o *OpenStack) gatherStoragePools() error {
+	if !o.services.ContainsService(volumeV2Service) {
+		return nil
+	}
+	results, err := schedulerstats.List(o.volume, &schedulerstats.ListOpts{Detail: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list storage pools: %v", err)
+	}
+	storagePools, err := schedulerstats.ExtractStoragePools(results)
+	if err != nil {
+		return fmt.Errorf("unable to extract storage pools: %v", err)
+	}
+	for _, storagePool := range storagePools {
+		o.storagePools[storagePool.Name] = storagePool
+	}
+	return nil
+}
+
+// accumulateIdentity accumulates statistics from the identity service.
+func (o *OpenStack) accumulateIdentity(acc telegraf.Accumulator) {
+	fields := fieldMap{
+		"projects": len(o.projects),
+	}
+	acc.AddFields("openstack_identity", fields, tagMap{})
+}
+
+// accumulateHypervisors accumulates statistics from hypervisors.
+func (o *OpenStack) accumulateHypervisors(acc telegraf.Accumulator) {
+	for _, hypervisor := range o.hypervisors {
+		tags := tagMap{
+			"name": hypervisor.HypervisorHostname,
```

Comment:
Some of the metrics are quite limited on the tags, would it be possible and make sense to have the server name and/or project here as well?  This way you could tie back the data to the containing object.

Reply:
Agree on this one being light on tags - and it is one of the things I have a patch about ready for when this lands. However - a "Hypervisor" doesn't belong to a project - it belongs to an AvailabilityZone - my patch gathers those and maps HV's to them, along with Hypervisor State/Status (which are management and operational states [ annoyingly named the opposite way around from what you'd expect ;-) ] )

Reply:
Fine with me to add this later. Maybe we could open an issue to not forget things?